caikit/runtime/health_probe.py

```diff
@@ -0,0 +1,267 @@
+    kubernetes pod where the server is also running.
+
+    Returns:
+        healthy (bool): True if all servers are healthy, False otherwise
+    """
+
+    # Get TLS key/cert files if possible
+    config = get_config()
+    tls_key = config.runtime.tls.server.key
+    tls_cert = config.runtime.tls.server.cert
+    client_ca = config.runtime.tls.client.cert
+    http_healthy, grpc_healthy = None, None
+
+    if config.runtime.http.enabled:
+        log.debug("Checking HTTP server health")
+        http_healthy = _http_health_probe(
+            config.runtime.http.port, tls_key, tls_cert, client_ca
+        )
+
+    if config.runtime.grpc.enabled:
+        log.debug("Checking gRPC server health")
+        grpc_healthy = _grpc_health_probe(
+            config.runtime.grpc.port, tls_key, tls_cert, client_ca
+        )
+
+    if False in [http_healthy, grpc_healthy]:
+        log.info(
+            "<RUN64273066I>",
+            "Server not healthy. HTTP: %s, gRPC: %s",
+            http_healthy,
+            grpc_healthy,
+        )
+        return False
+    return True
+
+
+## Implementation ##############################################################
+
+
+def _http_health_probe(
+    port: int,
+    tls_key: Optional[str],
+    tls_cert: Optional[str],
+    client_ca: Optional[str],
+) -> bool:
+    """Probe the http server
+
+    The implementation of this utility is a bit tricky because mTLS makes this
+    quite challenging. For insecure or TLS servers, we expect a valid healthy
+    response, but for mTLS servers, we may not have a valid key/cert pair that
+    the client can present to the server that is signed by the expected CA if
+    the trusted client CA does not match the one that signed the server's
+    key/cert pair.
+
+    The workaround for this is to detect SSLError and consider that to be a
+    passing health check. If the server is healthy enough to _reject_ bad SSL
+    requests, it's healthy enough to server good ones!
+
+    Args:
+        port (int): The port that the HTTP server is serving on
+        tls_key (Optional[str]): Body or path to the TLS key file if TLS/mTLS
+            enabled
+        tls_cert (Optional[str]): Body or path to the TLS cert file if TLS/mTLS
+            enabled
+        client_ca (Optional[str]): The client ca cert that the server is using
+            for mutual client auth
+
+    Returns:
+        healthy (bool): True if all servers are healthy, False otherwise
+    """
+    # NOTE: Local imports for optional dependency
+    with alog.ContextTimer(log.debug2, "Done with local grpc imports: "):
+
+        # Third Party
+        import requests  # pylint: disable=import-outside-toplevel
+
+        # Local
+        from .http_server import (  # pylint: disable=import-outside-toplevel
+            HEALTH_ENDPOINT,
+        )
+
+    # Requests requires that the TLS information be in files
+    with _tls_files(tls_key, tls_cert) as tls_files:
```

Comment:
is this leading to writing files with every call making it slow and should we cache the files?

Reply:
Oooooh, very good point

Reply:
As I think about this more, I don't think this is worth doing as it would require having non-temp write access and a configured location where the "cached" files get put. If this is too slow, users can simply use mounted files to begin with and not use inline TLS values.